### PR TITLE
Train / test sampling

### PR DESCRIPTION
**Pull request by DavidNemeskey**
Added a train/testset splitter + unified BatchWriter.

### CODE REVIEW

File: scripts/train_test_split.py [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+    if base_dir:
+        for relative_dir in relative_dirs:
+            (base_dir / relative_dir).mkdir(parents=True, exist_ok=True)
+
+
+def delete_empty_directories(base_dir: Path, relative_dirs: list[Path]):
+    if base_dir:
+        for relative_dir in relative_dirs[::-1]:
+            full_dir = base_dir / relative_dir
+            if sum(1 for _ in islice(full_dir.iterdir(), 1)) == 0:
```
Comment (acheronw) [inline scripts/train_test_split.py:74]: I think this is a simpler way to check if the dir is empty:

if not any(full_dir.iterdir())

File: scripts/train_test_split.py [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+            path.relative_to(args.input_dir)
+        )
+
+    logging.info(f'{len(relative_files)=}')
+    logging.info(f'{len(relative_dirs)=} {relative_dirs=}')
+    create_directories(args.output_dir, relative_dirs)
+    create_directories(args.excluded_dir, relative_dirs)
+
+    logging.info(f'Scheduled {len(relative_files)} files for sampling.')
+    random.seed(args.seed)
```
Comment (acheronw) [inline scripts/train_test_split.py:103]: There should be an option to use random.seed() without any explicit seeds given. Probably that should be the default case when the argument is not given, instead of 42.

File: cc_corpus/corpus.py [hunk 3]
```diff
@@ -14,6 +14,7 @@
 import re
 import shutil
 
+from cc_corpus.io import BatchWriter as BatchWriterBase
```
Comment (acheronw) [inline cc_corpus/corpus.py:17]: What does this cc_corpus.io refers to? I don't see any io.py files in the commit.

Comment (acheronw) [reply cc_corpus/corpus.py:17]: I absolutely support the idea of using a baseclass + inheritance to sort out the mess with the BatchWriters.

But:
1. The baseclass should be named as such, BatchWriterBase instead of renaming it during an import and then shadowing its original class name. 

2. I prefer the derived classes to have unique class names, like BatchWriterJsonFlagger or such. That is because sometimes we import them in other parts of the code, and it would greatly help code clarity.

3. There is a third BatchWriter class declared in deduplication.py, please refactor that as well to use inheritance!

**Review by acheronw — CHANGES_REQUESTED** [review]

File: scripts/train_test_split.py (hunk 2 above)
Comment (DavidNemeskey) [reply scripts/train_test_split.py:103]: I am not sure I want to allow a non-repeatable split.

File: cc_corpus/corpus.py (hunk 3 above)
Comment (DavidNemeskey) [reply cc_corpus/corpus.py]: 1. done
2-3. won't do for this PR, I would like to keep the "status quo" as much as possible. Let's return to this later. The third BatchWriter is a bit more complicated than the other two, so it might or might not fit into the "framework" anyway.

File: scripts/train_test_split.py (hunk 2 above)
Comment (acheronw) [reply scripts/train_test_split.py:102]: Ok.

File: cc_corpus/corpus.py (hunk 3 above)
Comment (acheronw) [reply cc_corpus/corpus.py]: Okay, but we should sort out this mess one day.

**Review by acheronw — APPROVED** [review]
